File: crates/delta-py/python/delta/nn/linear.py
from delta import Tensor
from delta.nn import Module, Parameter
import delta


class Linear(Module):

    def __init__(self, in_features: int, out_features: int, bias: bool = True) -> None:
        super().__init__()

        self.use_bias = bias
        if self.use_bias:
            in_features += 1
        self.weights = Parameter(delta.randn([in_features, out_features]))

    def forward(self, x: Tensor) -> Tensor:
        if self.use_bias:
            ones_shape = list(x.shape[:-1]) + [1]
            x = delta.cat([x, delta.ones(ones_shape)], dim=-1)
        return x @ self.weights.data

File: crates/delta-nn/src/linear.rs
use std::collections::HashMap;

use delta_tensor::{Tensor, device::Device, ivalue::IValue, linalg};

use crate::{Module, Parameter};

/// # `Linear` Layer
///
/// Contains of:
/// - weights
/// - bias
///
/// Linear layer performs: `x @ W + b`,
/// where:
/// - `x` is input
/// - `W` is weights
/// - `b` is bias
// #[derive(Module, Clone)]
pub struct Linear {
    // #[param]
    pub weights: Parameter,
}

impl Linear {
    pub fn new(mut in_features: usize, out_features: usize, bias: bool) -> Self {
        if bias {
            in_features += 1;
        }
        let _weights = delta_tensor::randn(&[in_features, out_features], Device::CPU);
        Self {
            weights: Parameter(_weights),
        }
    }
}

impl Module for Linear {
    fn module_name(&self) -> String {
        "Linear".to_owned()
    }

    fn parameters(&self) -> Vec<Parameter> {
        let parameters = vec![self.weights.clone()];
        parameters
    }

    fn forward(&self, args: Vec<IValue>, _kwargs: HashMap<String, IValue>) -> IValue {
        let x = match &args[0] {
            IValue::Tensor(t) => t.clone(),
            _ => panic!("Linear expects a Tensor as first argument"),
        };
        let dtype = x.dtype();
        let weights = self.weights.clone();
        let mut ones_shape = x.shape();
        let _ = ones_shape.pop();
        ones_shape.push(1);
        let x = Tensor::cat(
            &[x, delta_tensor::ones(&ones_shape, dtype, self.weights.device())],
            1,
        );
        let x = linalg::matmul(x, weights.0);
        IValue::Tensor(x)
    }
}
